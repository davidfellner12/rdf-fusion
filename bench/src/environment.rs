use crate::BenchmarkingOptions;
use crate::benchmarks::BenchmarkName;
use crate::prepare::{PrepRequirement, prepare_run_closure, prepare_run_command};
use crate::prepare::{ensure_file_download, prepare_file_download};
use anyhow::bail;
use std::path::{Path, PathBuf};
use std::sync::Mutex;

/// Represents a context used to execute benchmarks.
pub struct RdfFusionBenchContext {
    /// General options for the benchmarks.
    options: BenchmarkingOptions,
    /// The path to the data dir.
    data_dir: Mutex<PathBuf>,
    /// The path to the current directory. This might be different from the `data_dir` if a
    /// benchmark is running.
    results_dir: Mutex<PathBuf>,
}

impl RdfFusionBenchContext {
    /// Creates a new [RdfFusionBenchContext].
    pub fn new(
        options: BenchmarkingOptions,
        data_dir: PathBuf,
        results_dir: PathBuf,
    ) -> Self {
        Self {
            options,
            data_dir: Mutex::new(data_dir),
            results_dir: Mutex::new(results_dir),
        }
    }

    /// Returns the [BenchmarkingOptions] for this context.
    pub fn options(&self) -> &BenchmarkingOptions {
        &self.options
    }

    /// Resolves a relative path `file` against the data directory.
    pub fn join_data_dir(&self, file: &Path) -> anyhow::Result<PathBuf> {
        if !file.is_relative() {
            bail!("Only relative paths can be resolved.")
        }

        Ok(self.data_dir.lock().unwrap().join(file))
    }

    /// Creates a new folder in the results directory and uses it until [Self::pop_dir] is
    /// called.
    ///
    /// This can be used to create folder hierarchies to separate the results of different
    /// benchmarks.
    #[allow(clippy::create_dir)]
    pub fn push_dir(&self, dir: &str) -> anyhow::Result<()> {
        let mut data_dir = self.data_dir.lock().unwrap();
        let mut results_dir = self.results_dir.lock().unwrap();

        data_dir.push(dir);
        results_dir.push(dir);

        Ok(())
    }

    /// Pops the last directory from the stack.
    pub fn pop_dir(&self) {
        let mut data_dir = self.data_dir.lock().unwrap();
        let mut results_dir = self.results_dir.lock().unwrap();

        data_dir.pop();
        results_dir.pop();
    }

    /// Creates a new bencher and modifies the context for this benchmark.
    pub fn create_benchmark_context(
        &self,
        benchmark_name: BenchmarkName,
    ) -> anyhow::Result<BenchmarkContext<'_>> {
        self.push_dir(&benchmark_name.dir_name())?;
        Ok(BenchmarkContext {
            context: self,
            benchmark_name,
        })
    }
}

/// A benchmarker that can be used to execute benchmarks.
///
/// It holds a reference to the current context to store its results.
pub struct BenchmarkContext<'ctx> {
    /// Reference to the benchmarking context.
    context: &'ctx RdfFusionBenchContext,
    /// Name of the benchmark that is being executed.
    benchmark_name: BenchmarkName,
}

impl<'ctx> BenchmarkContext<'ctx> {
    /// Returns a reference to the benchmarking context.
    pub fn parent(&self) -> &RdfFusionBenchContext {
        self.context
    }

    /// Returns the name of the benchmark that is being executed.
    pub fn benchmark_name(&self) -> BenchmarkName {
        self.benchmark_name
    }

    /// Returns the path to the results directory of this benchmark.
    pub fn data_dir(&self) -> PathBuf {
        self.context.data_dir.lock().unwrap().clone()
    }

    /// Returns the path to the results directory of this benchmark.
    pub fn results_dir(&self) -> PathBuf {
        self.context.results_dir.lock().unwrap().clone()
    }

    /// Prepares the context such that `requirement` is fulfilled.
    pub async fn prepare_requirement(
        &self,
        requirement: PrepRequirement,
    ) -> anyhow::Result<()> {
        match requirement {
            PrepRequirement::FileDownload {
                url,
                file_name,
                action,
            } => prepare_file_download(self, url, file_name, action).await,
            PrepRequirement::RunClosure { execute, .. } => {
                prepare_run_closure(self, &execute)
            }
            PrepRequirement::RunCommand {
                workdir,
                program,
                args,
                ..
            } => {
                let workdir = self.context.join_data_dir(&workdir)?;
                prepare_run_command(&workdir, &program, &args)
            }
        }
    }

    /// Ensures that the `requirement` is fulfilled in this context.
    pub fn ensure_requirement(&self, requirement: PrepRequirement) -> anyhow::Result<()> {
        match requirement {
            PrepRequirement::FileDownload { file_name, .. } => {
                ensure_file_download(self, file_name.as_path())
            }
            PrepRequirement::RunClosure {
                check_requirement, ..
            }
            | PrepRequirement::RunCommand {
                check_requirement, ..
            } => check_requirement(self),
        }
    }
}

/// Pops the results directory from the context when the bencher is dropped.
impl Drop for BenchmarkContext<'_> {
    fn drop(&mut self) {
        self.context.pop_dir();
    }
}
