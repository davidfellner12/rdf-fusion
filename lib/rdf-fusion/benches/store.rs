#![allow(clippy::panic)]

use bzip2::read::MultiBzDecoder;
use codspeed_criterion_compat::{criterion_group, criterion_main, Criterion, Throughput};
use futures::StreamExt;
use oxrdfio::RdfFormat;
use rdf_fusion::store::Store;
use rdf_fusion_engine::results::QueryResults;
use rdf_fusion_engine::sparql::{Query, QueryOptions};
use reqwest::Url;
use std::fs::File;
use std::io::{Read, Write};
use std::path::Path;
use std::str;
use std::str::FromStr;
use tokio::runtime::Runtime;

fn store_load(c: &mut Criterion) {
    let data = read_bz2_data("https://zenodo.org/records/12663333/files/dataset-1000.nt.bz2");
    let mut group = c.benchmark_group("store load");
    group.throughput(Throughput::Bytes(data.len() as u64));
    group.sample_size(10);
    group.bench_function("load BSBM explore 1000 in memory", |b| {
        b.to_async(Runtime::new().unwrap()).iter(|| async {
            let store = Store::new();
            do_load(&store, &data).await;
        });
    });
}

async fn do_load(store: &Store, data: &[u8]) {
    store
        .load_from_reader(RdfFormat::NTriples, data)
        .await
        .unwrap();
}

fn store_query_and_update(c: &mut Criterion) {
    for data_size in [1_000, 5_000] {
        do_store_query_and_update(c, data_size)
    }
}

fn do_store_query_and_update(c: &mut Criterion, data_size: usize) {
    let data = read_bz2_data(&format!(
        "https://zenodo.org/records/12663333/files/dataset-{data_size}.nt.bz2"
    ));
    let explore_operations = bsbm_sparql_operation("exploreAndUpdate-1000.csv.bz2")
        .into_iter()
        .filter_map(|op| match op {
            RawOperation::Query(q) => {
                // TODO remove once describe is supported
                if q.contains("DESCRIBE") {
                    None
                } else {
                    Some(Operation::Query(Query::parse(&q, None).unwrap()))
                }
            }
            RawOperation::Update(_) => None,
        })
        .collect::<Vec<_>>();
    let explore_query_operations = explore_operations
        .iter()
        .filter(|o| matches!(o, Operation::Query(_)))
        .cloned()
        .collect::<Vec<_>>();
    let business_operations = bsbm_sparql_operation("businessIntelligence-1000.csv.bz2")
        .into_iter()
        .filter_map(|op| match op {
            RawOperation::Query(q) => {
                // TODO remove once describe is supported
                if q.contains("DESCRIBE") {
                    None
                } else {
                    Some(Operation::Query(
                        Query::parse(&q.replace('#', ""), None).unwrap(),
                    ))
                }
            }
            RawOperation::Update(_) => unreachable!(),
        })
        .collect::<Vec<_>>();

    let mut group = c.benchmark_group("store operations");
    group.sample_size(10);

    {
        let runtime = Runtime::new().unwrap();
        let memory_store = Store::new();
        runtime.block_on(do_load(&memory_store, &data));

        group.bench_function(format!("BSBM explore {data_size} query in memory"), |b| {
            b.to_async(Runtime::new().unwrap())
                .iter(|| run_operation(&memory_store, &explore_query_operations));
        });
        group.bench_function(
            format!("BSBM explore {data_size} queryAndUpdate in memory"),
            |b| {
                b.to_async(Runtime::new().unwrap())
                    .iter(|| run_operation(&memory_store, &explore_operations));
            },
        );
        group.bench_function(
            format!("BSBM business intelligence {data_size} in memory"),
            |b| {
                b.to_async(Runtime::new().unwrap())
                    .iter(|| run_operation(&memory_store, &business_operations));
            },
        );
    }
}

async fn run_operation(store: &Store, operations: &[Operation]) {
    let options = QueryOptions::default();
    for operation in operations {
        match operation {
            Operation::Query(q) => match store.query_opt(q.clone(), options.clone()).await.unwrap()
            {
                QueryResults::Boolean(_) => (),
                QueryResults::Solutions(mut s) => {
                    while let Some(s) = s.next().await {
                        s.unwrap();
                    }
                }
                QueryResults::Graph(mut g) => {
                    while let Some(t) = g.next().await {
                        t.unwrap();
                    }
                }
            },
            // Operation::Update(u) => store.update_opt(u.clone(), options.clone()).await.unwrap(),
        }
    }
}

criterion_group!(store, store_query_and_update, store_load);

criterion_main!(store);

fn read_bz2_data(url: &str) -> Vec<u8> {
    let url = Url::from_str(url).unwrap();
    let file_name = url.path().split('/').next_back().unwrap().to_owned();

    if !Path::new(&file_name).exists() {
        let client = reqwest::blocking::Client::new();
        let response = client.get(url.clone()).send().unwrap();
        assert!(response.status().is_success(), "{}", &url);

        File::create(&file_name)
            .unwrap()
            .write_all(&response.bytes().unwrap())
            .unwrap();
    }

    let mut buf = Vec::new();
    MultiBzDecoder::new(File::open(&file_name).unwrap())
        .read_to_end(&mut buf)
        .unwrap();
    buf
}

fn bsbm_sparql_operation(file_name: &str) -> Vec<RawOperation> {
    csv::Reader::from_reader(read_bz2_data(&format!("https://zenodo.org/records/12663333/files/{file_name}")).as_slice()).records()
        .collect::<Result<Vec<_>, _>>().unwrap()
        .into_iter()
        .rev()
        .take(300) // We take only 10 groups
        .map(|l| {
            match &l[1] {
                "query" => RawOperation::Query(l[2].into()),
                "update" => RawOperation::Update(l[2].into()),
                _ => panic!("Unexpected operation kind {}", &l[1]),
            }
        })
        .collect()
}

#[allow(dead_code)]
#[derive(Clone)]
enum RawOperation {
    Query(String),
    Update(String),
}

#[allow(clippy::large_enum_variant, clippy::allow_attributes)]
#[derive(Clone)]
enum Operation {
    Query(Query),
}
