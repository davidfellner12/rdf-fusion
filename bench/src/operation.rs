use crate::runs::BenchmarkRun;
use futures::StreamExt;
use rdf_fusion::store::Store;
use rdf_fusion::{Query, QueryExplanation, QueryOptions, QueryResults};

#[derive(Clone)]
pub enum SparqlRawOperation<QueryName> {
    Query(QueryName, String),
}

impl<QueryName: Clone> SparqlRawOperation<QueryName> {
    pub fn parse(&self) -> anyhow::Result<SparqlOperation<QueryName>> {
        match self {
            SparqlRawOperation::Query(query_name, query) => {
                let query = query.parse()?;
                Ok(SparqlOperation::Query(query_name.clone(), query))
            }
        }
    }
}

#[derive(Clone)]
pub enum SparqlOperation<QueryName> {
    Query(QueryName, Query),
}

impl<QueryName> SparqlOperation<QueryName> {
    pub fn query(&self) -> &Query {
        match self {
            SparqlOperation::Query(_, query) => query,
        }
    }

    pub async fn run(
        &self,
        store: &Store,
    ) -> anyhow::Result<(BenchmarkRun, QueryExplanation)> {
        let start = datafusion::common::instant::Instant::now();

        let options = QueryOptions::default();
        let explanation = match &self {
            SparqlOperation::Query(_, q) => {
                let (result, explanation) =
                    store.explain_query_opt(q.clone(), options.clone()).await?;
                match result {
                    QueryResults::Boolean(_) => (),
                    QueryResults::Solutions(s) => {
                        let mut stream = s.into_record_batch_stream()?;
                        while let Some(s) = stream.next().await {
                            s?;
                        }
                    }
                    QueryResults::Graph(mut g) => {
                        while let Some(t) = g.next().await {
                            t?;
                        }
                    }
                }
                explanation
            }
        };

        let duration = start.elapsed();
        Ok((BenchmarkRun { duration }, explanation))
    }
}

impl<QueryName: Clone> SparqlOperation<QueryName> {
    pub fn query_name(&self) -> QueryName {
        match self {
            SparqlOperation::Query(name, _) => name.clone(),
        }
    }
}
