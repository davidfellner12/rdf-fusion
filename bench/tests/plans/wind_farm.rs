use crate::plans::canonicalize_uuids;
use anyhow::Context;
use datafusion::physical_plan::displayable;
use datafusion::prelude::SessionConfig;
use insta::assert_snapshot;
use rdf_fusion::store::Store;
use rdf_fusion::{QueryExplanation, QueryOptions};
use rdf_fusion_bench::benchmarks::Benchmark;
use rdf_fusion_bench::benchmarks::windfarm::{
    NumTurbines, WindFarmBenchmark, WindFarmQueryName, get_wind_farm_raw_sparql_operation,
};
use rdf_fusion_bench::environment::{BenchmarkContext, RdfFusionBenchContext};
use rdf_fusion_bench::operation::SparqlRawOperation;
use std::path::PathBuf;

#[tokio::test]
pub async fn optimized_logical_plan_wind_farm() {
    for_all_explanations(|name, explanation| {
        assert_snapshot!(
            format!("{name} (Optimized)"),
            canonicalize_uuids(&explanation.optimized_logical_plan.to_string())
        )
    })
    .await;
}

#[tokio::test]
pub async fn execution_plan_wind_farm() {
    for_all_explanations(|name, explanation| {
        let string = displayable(explanation.execution_plan.as_ref())
            .indent(false)
            .to_string();
        assert_snapshot!(
            format!("{name} (Execution Plan)"),
            canonicalize_uuids(&string)
        )
    })
    .await;
}

async fn for_all_explanations(assertion: impl Fn(String, QueryExplanation) -> ()) {
    let benchmarking_context =
        RdfFusionBenchContext::new_for_criterion(PathBuf::from("./data"));

    // Load the benchmark data and set max query count to one.
    let benchmark = WindFarmBenchmark::new(NumTurbines::N4);
    let benchmark_name = benchmark.name();
    let benchmark_context = benchmarking_context
        .create_benchmark_context(benchmark_name)
        .unwrap();

    let store =
        Store::new_with_datafusion_config(SessionConfig::new().with_target_partitions(1));
    for query_name in WindFarmQueryName::list_queries() {
        let benchmark_name = format!("Wind Farm - {query_name}");
        let query = get_query_to_execute(&benchmark_context, query_name);

        let (_, explanation) = store
            .explain_query_opt(query.text(), QueryOptions::default())
            .await
            .unwrap();

        assertion(benchmark_name, explanation);
    }
}

fn get_query_to_execute(
    benchmark_context: &BenchmarkContext,
    query_name: WindFarmQueryName,
) -> SparqlRawOperation<WindFarmQueryName> {
    get_wind_farm_raw_sparql_operation(benchmark_context, query_name)
        .context("Could not list raw operations for Wind Farm benchmark. Have you prepared a wind-farm-4 dataset?")
        .unwrap()
}
