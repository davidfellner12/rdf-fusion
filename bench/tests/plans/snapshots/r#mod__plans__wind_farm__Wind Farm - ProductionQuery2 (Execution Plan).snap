---
source: bench/tests/plans/wind_farm.rs
expression: "&string"
---
ProjectionExec: expr=[ENC_PT(site_label@0) as site_label, ENC_PT(wtur_label@1) as wtur_label, ENC_PT(t@3) as t, ENC_PT(val@2) as val]
  CoalesceBatchesExec: target_batch_size=8192
    FilterExec: EBV(BOOLEAN_AS_TERM(EBV(BOOLEAN_AS_TERM(EBV(EQ(ENC_TV(wtur_label@3), ENC_TV({term_type:2,value:A1,data_type:http://www.w3.org/2001/XMLSchema#string,language_tag:}))) OR EBV(EQ(ENC_TV(wtur_label@3), ENC_TV({term_type:2,value:A2,data_type:http://www.w3.org/2001/XMLSchema#string,language_tag:}))) OR EBV(EQ(ENC_TV(wtur_label@3), ENC_TV({term_type:2,value:A3,data_type:http://www.w3.org/2001/XMLSchema#string,language_tag:}))) OR EBV(EQ(ENC_TV(wtur_label@3), ENC_TV({term_type:2,value:A4,data_type:http://www.w3.org/2001/XMLSchema#string,language_tag:}))) OR EBV(EQ(ENC_TV(wtur_label@3), ENC_TV({term_type:2,value:A5,data_type:http://www.w3.org/2001/XMLSchema#string,language_tag:}))) OR EBV(EQ(ENC_TV(wtur_label@3), ENC_TV({term_type:2,value:A6,data_type:http://www.w3.org/2001/XMLSchema#string,language_tag:}))) OR EBV(EQ(ENC_TV(wtur_label@3), ENC_TV({term_type:2,value:A7,data_type:http://www.w3.org/2001/XMLSchema#string,language_tag:}))) OR EBV(EQ(ENC_TV(wtur_label@3), ENC_TV({term_type:2,value:A8,data_type:http://www.w3.org/2001/XMLSchema#string,language_tag:}))) OR EBV(EQ(ENC_TV(wtur_label@3), ENC_TV({term_type:2,value:A9,data_type:http://www.w3.org/2001/XMLSchema#string,language_tag:}))) OR EBV(EQ(ENC_TV(wtur_label@3), ENC_TV({term_type:2,value:A10,data_type:http://www.w3.org/2001/XMLSchema#string,language_tag:}))))) AND EBV(EQ(ENC_TV(site_label@1), ENC_TV({term_type:2,value:Wind Mountain,data_type:http://www.w3.org/2001/XMLSchema#string,language_tag:}))))), projection=[site_label@1, wtur_label@3, val@11, t@12]
      CoalesceBatchesExec: target_batch_size=8192
        HashJoinExec: mode=CollectLeft, join_type=Inner, on=[(dp@10, dp@0)], projection=[site@0, site_label@1, wtur_asp@2, wtur_label@3, wtur@4, gensys_asp@5, gensys@6, generator_asp@7, generator@8, ts@9, dp@10, val@11, t@13]
          CoalesceBatchesExec: target_batch_size=8192
            HashJoinExec: mode=CollectLeft, join_type=Inner, on=[(dp@10, dp@0)], projection=[site@0, site_label@1, wtur_asp@2, wtur_label@3, wtur@4, gensys_asp@5, gensys@6, generator_asp@7, generator@8, ts@9, dp@10, val@12]
              CoalesceBatchesExec: target_batch_size=8192
                HashJoinExec: mode=CollectLeft, join_type=Inner, on=[(ts@9, ts@0)], projection=[site@0, site_label@1, wtur_asp@2, wtur_label@3, wtur@4, gensys_asp@5, gensys@6, generator_asp@7, generator@8, ts@9, dp@11]
                  CoalesceBatchesExec: target_batch_size=8192
                    HashJoinExec: mode=CollectLeft, join_type=Inner, on=[(ts@9, ts@0)], projection=[site@0, site_label@1, wtur_asp@2, wtur_label@3, wtur@4, gensys_asp@5, gensys@6, generator_asp@7, generator@8, ts@9]
                      CoalesceBatchesExec: target_batch_size=8192
                        HashJoinExec: mode=CollectLeft, join_type=Inner, on=[(generator@8, generator@0)], projection=[site@0, site_label@1, wtur_asp@2, wtur_label@3, wtur@4, gensys_asp@5, gensys@6, generator_asp@7, generator@8, ts@10]
                          CoalesceBatchesExec: target_batch_size=8192
                            HashJoinExec: mode=CollectLeft, join_type=Inner, on=[(generator@8, generator@0)], projection=[site@0, site_label@1, wtur_asp@2, wtur_label@3, wtur@4, gensys_asp@5, gensys@6, generator_asp@7, generator@8]
                              CoalesceBatchesExec: target_batch_size=8192
                                HashJoinExec: mode=CollectLeft, join_type=Inner, on=[(generator_asp@7, generator_asp@1)], projection=[site@0, site_label@1, wtur_asp@2, wtur_label@3, wtur@4, gensys_asp@5, gensys@6, generator_asp@7, generator@8]
                                  CoalesceBatchesExec: target_batch_size=8192
                                    HashJoinExec: mode=CollectLeft, join_type=Inner, on=[(gensys@6, gensys@0)], projection=[site@0, site_label@1, wtur_asp@2, wtur_label@3, wtur@4, gensys_asp@5, gensys@6, generator_asp@8]
                                      CoalesceBatchesExec: target_batch_size=8192
                                        HashJoinExec: mode=CollectLeft, join_type=Inner, on=[(gensys@6, gensys@0)], projection=[site@0, site_label@1, wtur_asp@2, wtur_label@3, wtur@4, gensys_asp@5, gensys@6]
                                          CoalesceBatchesExec: target_batch_size=8192
                                            HashJoinExec: mode=CollectLeft, join_type=Inner, on=[(gensys_asp@5, gensys_asp@1)], projection=[site@0, site_label@1, wtur_asp@2, wtur_label@3, wtur@4, gensys_asp@5, gensys@6]
                                              CoalesceBatchesExec: target_batch_size=8192
                                                HashJoinExec: mode=CollectLeft, join_type=Inner, on=[(wtur@4, wtur@0)], projection=[site@0, site_label@1, wtur_asp@2, wtur_label@3, wtur@4, gensys_asp@5]
                                                  CoalesceBatchesExec: target_batch_size=8192
                                                    HashJoinExec: mode=CollectLeft, join_type=Inner, on=[(wtur@4, wtur@0)], projection=[site@0, site_label@1, wtur_asp@2, wtur_label@3, wtur@4, gensys_asp@6]
                                                      CoalesceBatchesExec: target_batch_size=8192
                                                        HashJoinExec: mode=CollectLeft, join_type=Inner, on=[(wtur_asp@2, wtur_asp@1)], projection=[site@0, site_label@1, wtur_asp@2, wtur_label@3, wtur@4]
                                                          CoalesceBatchesExec: target_batch_size=8192
                                                            HashJoinExec: mode=CollectLeft, join_type=Inner, on=[(wtur_asp@2, wtur_asp@0)], projection=[site@0, site_label@1, wtur_asp@2, wtur_label@4]
                                                              CoalesceBatchesExec: target_batch_size=8192
                                                                HashJoinExec: mode=CollectLeft, join_type=Inner, on=[(site@0, site@0)], projection=[site@0, site_label@1, wtur_asp@3]
                                                                  CoalesceBatchesExec: target_batch_size=8192
                                                                    HashJoinExec: mode=CollectLeft, join_type=Inner, on=[(site@0, site@0)], projection=[site@0, site_label@2]
                                                                      DataSourceExec: [GPOS] subject=?site, predicate=<http://www.w3.org/1999/02/22-rdf-syntax-ns#type>, object=<https://github.com/magbak/chrontext/rds_power#Site>
                                                                      DataSourceExec: [GPOS] subject=?site, predicate=<http://www.w3.org/2000/01/rdf-schema#label>, object=?site_label, additional_filters=[DynamicFilterPhysicalExpr [ true ]]
                                                                  DataSourceExec: [GPOS] subject=?site, predicate=<https://github.com/magbak/chrontext/rds_power#hasFunctionalAspect>, object=?wtur_asp, additional_filters=[DynamicFilterPhysicalExpr [ true ]]
                                                              DataSourceExec: [GPOS] subject=?wtur_asp, predicate=<http://www.w3.org/2000/01/rdf-schema#label>, object=?wtur_label, additional_filters=[DynamicFilterPhysicalExpr [ true ]]
                                                          DataSourceExec: [GPOS] subject=?wtur, predicate=<https://github.com/magbak/chrontext/rds_power#hasFunctionalAspectNode>, object=?wtur_asp, additional_filters=[DynamicFilterPhysicalExpr [ true ]]
                                                      DataSourceExec: [GPOS] subject=?wtur, predicate=<https://github.com/magbak/chrontext/rds_power#hasFunctionalAspect>, object=?gensys_asp, additional_filters=[DynamicFilterPhysicalExpr [ true ]]
                                                  DataSourceExec: [GPOS] subject=?wtur, predicate=<http://www.w3.org/1999/02/22-rdf-syntax-ns#type>, object=<https://github.com/magbak/chrontext/rds_power#A>, additional_filters=[DynamicFilterPhysicalExpr [ true ]]
                                              DataSourceExec: [GPOS] subject=?gensys, predicate=<https://github.com/magbak/chrontext/rds_power#hasFunctionalAspectNode>, object=?gensys_asp, additional_filters=[DynamicFilterPhysicalExpr [ true ]]
                                          DataSourceExec: [GPOS] subject=?gensys, predicate=<http://www.w3.org/1999/02/22-rdf-syntax-ns#type>, object=<https://github.com/magbak/chrontext/rds_power#RA>, additional_filters=[DynamicFilterPhysicalExpr [ true ]]
                                      DataSourceExec: [GPOS] subject=?gensys, predicate=<https://github.com/magbak/chrontext/rds_power#hasFunctionalAspect>, object=?generator_asp, additional_filters=[DynamicFilterPhysicalExpr [ true ]]
                                  DataSourceExec: [GPOS] subject=?generator, predicate=<https://github.com/magbak/chrontext/rds_power#hasFunctionalAspectNode>, object=?generator_asp, additional_filters=[DynamicFilterPhysicalExpr [ true ]]
                              DataSourceExec: [GPOS] subject=?generator, predicate=<http://www.w3.org/1999/02/22-rdf-syntax-ns#type>, object=<https://github.com/magbak/chrontext/rds_power#GAA>, additional_filters=[DynamicFilterPhysicalExpr [ true ]]
                          DataSourceExec: [GPOS] subject=?generator, predicate=<https://github.com/magbak/chrontext#hasTimeseries>, object=?ts, additional_filters=[DynamicFilterPhysicalExpr [ true ]]
                      DataSourceExec: [GPOS] subject=?ts, predicate=<http://www.w3.org/2000/01/rdf-schema#label>, object="Production", additional_filters=[DynamicFilterPhysicalExpr [ true ]]
                  DataSourceExec: [GPOS] subject=?ts, predicate=<https://github.com/magbak/chrontext#hasDataPoint>, object=?dp, additional_filters=[DynamicFilterPhysicalExpr [ true ]]
              DataSourceExec: [GPOS] subject=?dp, predicate=<https://github.com/magbak/chrontext#hasValue>, object=?val, additional_filters=[DynamicFilterPhysicalExpr [ true ]]
          DataSourceExec: [GPOS] subject=?dp, predicate=<https://github.com/magbak/chrontext#hasTimestamp>, object=?t, additional_filters=[DynamicFilterPhysicalExpr [ true ]]
